File: solutions/207.course-schedule.java
/*
 * @lc app=leetcode id=207 lang=java
 *
 * [207] Course Schedule
 */

// @lc code=start

import java.util.ArrayList;
import java.util.List;

class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        List<List<Integer>> courses = new ArrayList<>(numCourses);
        for(int i = 0; i < numCourses; i++) {
            courses.add(new ArrayList<>());
        }
        for (int[] pair : prerequisites) {
            courses.get(pair[0]).add(pair[1]);
         }

         int[] visited = new int[numCourses];
         for(int i = 0; i < numCourses; i++) {
            if (visited[i] == 0 && hasCycle(i, courses, visited)) {
                return false;
            }
         }

         return true;
    }

    private boolean hasCycle(int index, List<List<Integer>> courses, int[] visited) {
        if (visited[index] == 1) {
            return true;    // visiting. cycle detected
        }
        if (visited[index] == 2) {
            return false;   // visited. no need to check second time
        }

        visited[index] = 1;
        for(int child : courses.get(index)) {
            if (hasCycle(child, courses, visited)) {
                return true;    // return if found cycle 
            }
        }
        visited[index] = 2;     // No need to check children second time from another starting point

        return false;
    }
}
// @lc code=end

File: solutions/207.course-schedule.py
#
# @lc app=leetcode id=207 lang=python3
#
# [207] Course Schedule
#

# @lc code=start
from typing import List


class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        courses = [[] for _ in range(numCourses)]
        for pair in prerequisites:
            courses[pair[0]].append(pair[1])

        def has_cycle(index):
            if visited[index] == 1:
                return True     # visiting. cycle detected
            if visited[index] == 2:
                return False    # visited. no need to check child second time

            visited[index] = 1
            for child in courses[index]:
                if has_cycle(child):
                    return True
            visited[index] = 2

            return False

        visited = [0] * numCourses
        for index in range(numCourses):
            if visited[index] == 0 and has_cycle(index):
                return False
            
        return True

# @lc code=end

